Backstop is working locally

alexjanousekGSA opened the pull request:
<!--
Not sure what you should include or write in a pull request?  Please read the
[pull request documentation in our docs!](https://github.com/GSA/notifications-api/blob/main/docs/all.md#pull-requests)
-->

*A note to PR reviewers: it may be helpful to review our [code review documentation](https://github.com/GSA/notifications-api/blob/main/docs/all.md#code-reviews) to know what to keep in mind while reviewing pull requests.*

## Description

Please enter a detailed description here.

## TODO (optional)

* [ ] TODO item 1
* [ ] TODO item 2
* [ ] TODO item ...

## Security Considerations

* Consideration 1
* Consideration 2
* Consideration ...

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -139,34 +140,22 @@ def verify_email(user, redirect_url):
-        stre = stre.replace(" ", "_")
-        # Trying to get a message back to playwright somehow since we can't see the admin logs
-        return redirect(url_for(f"https://{stre}"))
+    current_app.logger.warning("E2E TESTS ARE ENABLED.")
+    current_app.logger.warning(
+        "If you are getting a 404 on signin, comment out E2E vars in .env file!"
+    )
+    user = user_api_client.get_user_by_email(os.getenv("NOTIFY_E2E_TEST_EMAIL"))
+    activate_user(user["id"])
+
+    if redirect_url and is_safe_redirect_url(redirect_url):
+        return redirect(redirect_url)
+
+    return redirect(url_for("main.show_accounts_or_dashboard"))
```
I think this whole section might need to be updated again as it is undoing the additional error handling that was recently put into place.

@terrazoon could you please take a look here and double check?

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -178,6 +167,11 @@ def sign_in():
     ):
         return email_verify_template
 
+    redirect_url = request.args.get("next")
+
+    if os.getenv("NOTIFY_E2E_TEST_EMAIL"):
+        return _handle_e2e_tests(redirect_url)
+
```
Same with this change - checking with @terrazoon to verify!

ccostino commented on backstop.config.js:
```diff
@@ -0,0 +1,56 @@
+      height: 768,
+    },
+  ],
+  scenarios: [
```
Very cool!  Is this the part of the config that we'd add more paths to check to have tested in the future?

ccostino commented on docs/backstop.md:
````diff
@@ -0,0 +1,65 @@
+First, install Gulp globally if its not installed:
+
+```
+npm install -g gulp
````
Is this necessary to install globally, or would having `gulp` available in the project itself be enough?

alexjanousekGSA replied:
We just need to be able to run the command so having it in the project is enough if the command still works, ill test it.

ccostino commented on docs/backstop.md:
```diff
@@ -0,0 +1,65 @@
+
+The entry point for BackstopJS configuration can be found in the `backstop.config.js` file.
+
+To add a new feature or page to be tested, modify the `scenarios` array. Refer to the [official documentation](https://github.com/garris/BackstopJS) for details on the structure of the object.
```
Ah yes, this answers my question from before - thanks for calling it out here! 🙂

ccostino reviewed (changes requested):
Awesome, thanks @alexjanousekGSA!

I'm assuming the addition of the images are necessary to help establish a baseline and that more would be included in the future, is that correct?

There were a couple of things that caught my eye with the changes to the sign-in code, so I'm asking @terrazoon to help take a look too.

terrazoon reviewed (approved):
It looks like it is trying to override the most recent signin changes, which were pretty trivial.  Probably synching to main fixes it.   I didn't see anything worrisome.

ccostino commented on docs/backstop.md:
```diff
@@ -0,0 +1,65 @@
+Before running BackstopJS, ensure the following:
+
+- Both the API and Admin projects must be running.
+- In the .env file of the API project, make sure the following environment variables are uncommented and set to matching values:
+  - `NOTIFY_E2E_TEST_EMAIL`
+  - `NOTIFY_E2E_TEST_PASSWORD`
```
These need to be set for _both_ the API and the Admin!  Otherwise the tests won't work.

In other words, the Admin site needs to be running in E2E test mode, which is enabled by having both of these values set (and matching what's been defined for the API).

ccostino replied:
```suggestion
- In the .env file of both the API and Admin projects, make sure the following environment variables are uncommented and set to matching values:
  - `NOTIFY_E2E_TEST_EMAIL`
  - `NOTIFY_E2E_TEST_PASSWORD`
```

heyitsmebev reviewed (commented):
I just did a git pull on the branch and went through the steps in the documentation again. I'm noticing an error after running gulp backstopTest, but only when there's a failure in the reports. I'm also seeing that the reports are inconsistent when showing failures. For example, if I make a change to the features.html page, sometimes the report will show an error, and other times it won’t. Do Jinja blocks affect what is detected and what isn't? I've noticed more consistent failure reports when editing the body of the content versus the page title.

here's the error i get
```
[10:11:51] Error: Command failed: npx backstop test --configPath=backstop.config.js

    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at ChildProcess.exithandler (node:child_process:421:12)
    at ChildProcess.emit (node:events:520:28)
    at ChildProcess.emit (node:domain:551:15)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5)
    at Process.callbackTrampoline (node:internal/async_hooks:130:17)
```

alexjanousekGSA commented:
Update: Merge conflicts resolve. Focus was on the sign_in page for this. I tested and verified all of my changes work without removing or altering any of the changes in the main branch. 

I updated the readme to not reccomend installing gulp globally, instead it is now within the project and updated it to remind to run `make bootstrap` before using. 

Additionally, I removed one missed accounts url that shouldn't be included because its unique to each individual. I also added a slight delay to each screenshot which might help with any discrepancies with inconsistency. If that didn't work then we can change the mismatch percentage. However the higher we go with that, the less it will catch very small things. 

I also added a .prettierrc in which we can start adding code formatting standards we recommend. You can tell that my formatter updated one file by adding a trailing comma (which I personally like because it reduces line changes when altering the file for PR's), but it seemed as if the project already adopted single quotes, so I added that to the .prettierrc. In the future we could also add prettier to format as a github hook, only on changed files to minimize code changes.

alexjanousekGSA commented:
> Awesome, thanks @alexjanousekGSA!
> 
> I'm assuming the addition of the images are necessary to help establish a baseline and that more would be included in the future, is that correct?
> 
> There were a couple of things that caught my eye with the changes to the sign-in code, so I'm asking @terrazoon to help take a look too.

Also yes, as we think of new use cases (aka scenarios) then we will re-run the reference command. Typically that command will be run a lot, but with the expectation that we know there are changes and it highlights. If I was going to update an existing component. I would want to run the test, make sure it fails on what I expect it to fail on, then run the reference. That way we can make sure other pages are not affected. This will come handy when we remove UK uI from it.

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -97,11 +97,10 @@ def _get_access_token(code, state):  # pragma: no cover
-        # Capture the response json here so it hopefully shows up in error reports
         current_app.logger.exception(
-            f"Error when getting access token {response.json()} #notify-admin-1505"
+            f"Error when getting access token {response_json()} #notify-admin-1505"
```
It looks like these are still being changed, but I think it was intentional to refer to `response.json()` in this context because there was a key being looked for that didn't exist.  The error in this case is showing the original response content so that someone could verify this in the logs and check to see what's happening or missing.

```suggestion
            f"Error when getting access token {response.json()} #notify-admin-1505"
```

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -97,11 +97,10 @@ def _get_access_token(code, state):  # pragma: no cover
+            f"Error when getting access token {response_json()} #notify-admin-1505"
         )
-        raise KeyError(f"'access_token' {response.json()}") from e
+        raise KeyError(f"'access_token' {response_json()}") from e
```
Same with this log statement regarding the intent behind including the original response object:

```suggestion
        raise KeyError(f"'access_token' {response.json()}") from e
```

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -157,8 +156,9 @@ def _do_login_dot_gov():  # $ pragma: no cover
             activate_user(usr.id)
         except BaseException as be:  # noqa B036
-            current_app.logger.exception(f"Error signing in: {be} #notify-admin-1505 ")
+            current_app.logger.error(f"Error signing in: {be} #notify-admin-1505 ")
```
Nice catch, it looks like this one was missed from the others that were flipped to `error`!

terrazoon replied:
If you change this to exception, you don't need the {be}.

ccostino replied:
@xlorepdarkhelm changed the other lines to be `error` instead of `exception` so I'm assuming this one was meant to be changed as well and was just missed.  Thanks for the heads up though with the use of `exception`!

ccostino commented on app/main/views/sign_in.py:
```diff
@@ -213,6 +214,11 @@ def sign_in():  # pragma: no cover
     ):
         return email_verify_template
 
+    redirect_url = request.args.get("next")
+
+    if os.getenv("NOTIFY_E2E_TEST_EMAIL"):
+        return _handle_e2e_tests(redirect_url)
+
```
Double checking again with @terrazoon on this - should this have moved like, or does it need to go back to where it originally was?

terrazoon replied:
we want to skip all login.gov processing, so I moved the check for the email to the top of the method.  It can be verified easily if anything breaks:

1. Do e2e tests still pass?
2. Can you still log in?

ccostino replied:
@alexjanousekGSA were things not working with this code at the top of the method?  Or did that update somehow not roll into your other changes here with your last sync of `main` in this branch?

alexjanousekGSA replied:
Okay, somehow it didn't make its way back to the top of the method. I moved it back up there and tested it and it seems to all work fine!

ccostino reviewed (changes requested):
Thank you, @alexjanousekGSA!

There were a few more things I wanted to double check on, one of which I tagged @terrazoon on again as well.  After these are addressed I think we'll be good to go!

ccostino reviewed (approved):
Thank you, @alexjanousekGSA, I know this took some work! 🎉
